First 7 pg_xxxx types ready in draft.

driekus77 opened the pull request:
Fixed building and running unit tests on macbook m3 (arm64/silicon) Added types.hpp from ProgreSQL source pg_type.h.

driekus77 committed:
[WIP] Minor fixes in CMakeLists.txt and adding types.hpp
Fixed building and running unit tests on macbook m3 (arm64/silicon)
Added types.hpp from ProgreSQL source pg_type.h.

anarthal commented on CMakeLists.txt:
```diff
@@ -6,6 +6,26 @@
+            OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    # Specify the C and CXX compilers
+    set(CMAKE_C_COMPILER clang)
```
CMakeLists.txt should never set any of the `CMAKE_XXXXX` variables, including `CMAKE_C_COMPILER`, `CMAKE_CXX_COMPILER`, `CMAKE_PREFIX_PATH` or `CMAKE_BUILD_TYPE`. These are supposed to be set by the end user, via the command line or toolchain files (i.e. `-DCMAKE_BUILD_TYPE=Debug`).

driekus77 committed:
[WIP] Fixing review comments
Prototype setup using template types: See binary_writer.hpp

driekus77 committed:
[WIP] Deleted types.hpp in src folder and moved the fixed version to include/nativepg/types

driekus77 commented on src/types.hpp:
```diff
@@ -0,0 +1,150 @@
+
+/// PostgreSQL built-in type OIDs.
+/// Source: pg_type.h (PostgreSQL source)
+enum TypeOid : uint32_t
```
Fixed this in the include/nativepg/types/types.hpp version.

anarthal commented:
Actually, all types support text encoding, but some of them might not support binary, is that right? If that's the case, I think this could be useful:

```cpp
template <class T>
struct supports_binary : std::false_type {}; // Types don't support binary by default

template <>
struct supports_binary<int>: std::true_type {}; // specialize for types that do
```

driekus77 committed:
[WIP] UML design image showing responsibilities per type
And sketch implementation for basic_pg_value(s) and pg_type_traits

driekus77 committed:
[WIP] Trying to fix design image

driekus77 committed:
[WIP] Trying to fix design image using markdown file

driekus77 committed:
First 7 pg_xxx types ready: pg_bool, pg_int2, pg_int4, pg_float4, pg_float8, pg_text, and pg_uuid.
Constructing it from pg_type_traits and basic_pg_value
Created unit tests for them
Note: binary_writer is not used yet. Maybe become obsolete...

driekus77 commented:
Note: I prefer these 7 types being reviewed first. When they are oke I will create the rest of the types.
But don't hesitate to make a different choice.

anarthal commented on include/nativepg/types/basic_pg_value.hpp:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    // ----- runtime text encode/decode -----
+    std::string encode_text() const
```
Serialization methods need to accept the target object (preferentially a `vector<unsigned char>` or `vector<byte>`) by lvalue reference. Otherwise, you're copying the serialized data twice (once into a temporary buffer, then again into the request buffer). Pass-by-reference also allows for memory re-use.

My long-term plan is making this buffer reference generic, but let's go simple for now.

anarthal replied:
Hm, not quite. Something like

```cpp
struct basic_pg_value {
    void encode_text(std::vector<std::byte>& to) const;
};
```

Or, if you were using `std::string` as the target container for serialization:

```cpp
struct basic_pg_value {
    void encode_text(std::string& to) const;
};
```

The idea being that the target container is taken by reference rather than returned, because this way you can re-use the same storage multiple times. I prefer the `vector<byte>&` version because it's the container type that `protocol/` classes use for serialization, so that'd make the interaction smoother.

anarthal commented on include/nativepg/types/pg_uuid.hpp:
```diff
@@ -0,0 +1,158 @@
+namespace types {
+
+template <>
+struct pg_type_traits<std::array<std::uint8_t, 16>, pg_oid_type::uuid>
```
Might it make sense to use Boost.Uuid to represent this?

anarthal commented on include/nativepg/types/pg_int4.hpp:
```diff
@@ -0,0 +1,91 @@
+
+    static value_type decode_text(std::string_view sv)
+    {
+        return static_cast<std::int32_t>(std::stoi(std::string{sv}));
```
This is locale-sensitive and allocates. Use `std::from_chars`, instead.

anarthal commented on include/nativepg/types/pg_type_traits.hpp:
```diff
@@ -0,0 +1,67 @@
+namespace types {
+
+
+template <typename T, pg_oid_type Oid>
```
Why does the traits type need the Oid? Isn't a type like `std::int32_t` always associated to `pg_oid_type::int4`, for instance?

anarthal commented on include/nativepg/types/pg_int4.hpp:
```diff
@@ -0,0 +1,91 @@
+
+
+template <>
+struct pg_type_traits<std::int32_t, pg_oid_type::int4>
```
Suggestion for integers: you can factorize common code into a `detail::pg_type_traits<IntType>`, then define

```cpp
template <> struct pg_type_traits<std::int32_t, pg_oid_type::int4> : detail::pg_type_traits<std::int32_t, pg_oid_type::int4> {};
```
```

driekus77 commented on include/nativepg/types/pg_type_traits.hpp:
```diff
@@ -0,0 +1,67 @@
+
+
+template <typename T, pg_oid_type Oid>
+struct pg_type_traits
```
As above already notice: Maybe a good idea to let the user add/register a type mapper / traits? for every request / query?

anarthal replied:
Does the user get to choose the OID of the types they add to the database? In other words, is the OID for custom types available at compile-time, or only at runtime? Depending on this, we can make the register compile-time, based on customization points or ADL; or runtime.

anarthal commented on include/nativepg/types/basic_pg_value.hpp:
```diff
@@ -0,0 +1,172 @@
+// -----------------------------------------------------------------------------
+
+template <typename Traits>
+struct basic_pg_value
```
In Boost.JSON and Boost.MySQL we use reflection (via Boost.Describe or Boost.PFR - people love this last one) to perform the mapping. Starting in C++26 we will be able to do it without any library. These libs give you a compile-time list of member names and pointers given a type. I suggest doing similar here.

I don't know how the Entity Framework works unfortunately. Could you elaborate how this would go in C++?

driekus77 committed:
Fixed review comments
- No Exceptions but now using boost::system::error_code
- Using 2 new concepts: readable_traits and writeable_traits in types
- Added params.hpp for secure providing, through environment variables, the connection parameters in key/value pair(s).
- Removed basic_pg_value and binary_writer
- Added TYPES_MAPPING.md docu for overview of type mapping.

driekus77 commented:
@anarthal Sorry for late response. I was very busy with other things.

I think i've managed to come near your UML as far as I can. 

- Now using Boost Endian and Boost Uuid libs.
- Replaced exceptions with boost::system::error_code
- Introduced 2 new concepts in front of pg_type_trait: readable_trait and writeable_trait.
- Removed basic_pg_value and binary_writer which are now obsolete.
- Fixed the unit-tests acoordingly.

driekus77 committed:
Fixed incorrect unit tests
And removed "default" C++ to OID mapping from pg_type_traits and alike.

driekus77 commented:
@anarthal You did great work with this lib. 
And so much work that I'm in doubt if this PR is still work in the good direction? Should I close it? 
Do you still need help with the types? Or in any other way?

Kind regards,
Henry

anarthal commented:
Welcome. Sorry for getting ahead of you with the first types, I needed them to get the first things working. I think we can close the PR for now.

How about implementing date/time parsing? At the end, the objective would be being able to use C++ datetime types (e.g. `std::chrono::sys_time`) in the custom structs. You'd need add new specializations to `field_traits.hpp`. Let me know if you need any help.

Thanks,
Ruben.
